crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -0,0 +1,787 @@
+                            WriteOperation::Insert(old_value.into()),
+                        );
+                    }
+                    (Some(old_value), WriteOperation::Insert(new_value)) => {
+                        if old_value.as_slice() != new_value.as_slice() {
+                            entry.insert(
+                                key.clone(),
+                                WriteOperation::Insert(old_value.into()),
+                            );
+                        }
+                    }
+                }
+            }
+        }
+        Ok(reverse_changes)
+    }
+
+    pub fn latest_view(&self) -> StorageResult<RocksDb<Description>> {
+        let db_view = self.db.create_snapshot();
+        Ok(db_view)
+    }
+
+    pub fn create_view_at(
+        &self,
+        height: &Description::Height,
+    ) -> StorageResult<ViewAtHeight<Description>> {
+        let latest_view = self.latest_view()?;
+
+        // Each height stores reverse modification caused by the corresponding
+        // block at the same height. Applying reverse changes at height `X`
+        // gives us a state at height `X - 1`. If we want a state at height `X`,
+        // we need to apply all modifications up to `X + 1`.
+        let rollback_height = height.as_u64().saturating_add(1);
+
+        let Some(oldest_height) = self.oldest_changes_height()? else {
+            return Err(DatabaseError::NoHistoryIsAvailable.into());
+        };
+
+        if rollback_height < oldest_height {
+            return Err(DatabaseError::NoHistoryForRequestedHeight {
+                requested_height: height.as_u64(),
+                oldest_available_height: oldest_height.saturating_sub(1),
+            }
+            .into());
+        }
+
+        let history_view = self.history.create_snapshot();
+
+        Ok(ViewAtHeight::new(
+            rollback_height,
+            latest_view,
+            history_view,
+        ))
+    }
+
+    fn store_modifications_history(
+        &self,
+        height: &Description::Height,
+        changes: &Changes,
+    ) -> StorageResult<()> {
+        if self.state_rewind_policy == StateRewindPolicy::NoRewind {
+            return Ok(());
+        }
+
+        let height_u64 = height.as_u64();
+        let mut storage_transaction = self.history.read_transaction();
+
+        cleanup_old_changes(
+            &height_u64,
+            &mut storage_transaction,
+            &self.state_rewind_policy,
+        )?;
+
+        let reverse_changes = self.reverse_history_changes(changes)?;
+        let old_changes = storage_transaction
+            .storage_as_mut::<ModificationsHistory<Description>>()
+            .insert(&height_u64, &reverse_changes)?;
+
+        // The existence of old changes at the current height means we
+        // try to override the current height(maybe because of crash).
+        // So, we need to remove all historical modifications from the past.
+        if let Some(old_changes) = old_changes {
```

Comment:
We should probably do some kind of consistency check like this when the historical database is started. Otherwise, it may be possible to generate views at heights which shouldn't exist.